devel/design/networking-api.md

```diff
@@ -0,0 +1,509 @@
+
+#### Ansible
+Ansible represents in this context a client of RHEL, through which one can configure its networking configuration.
+
+Atomicity is embedded into nmstate, therefore, no special treatment is required in that area as well.
+
+#### Other Management and Control clients
+When monitoring capabilities are required or Ansible is not the best fit for a solution, other management and controlling clients will need to access the RHEL node directly.
+
+Such clients, similar to Ansible, should support the host network schema and the communication protocol. The clients may also provide cluster level transaction.
+
+### Deployment
+Containerizing the components for deployment is recommended.
+
+Most networking actions require privilege access to the system. In practice, only NetworkManager and the other backends require such access, all the rest should be fine with unprivileged access. It is therefore suggested to use a linker between the two domains (NM on the host root namespace and all the other components in a container).
+NMLIB or D-BUS directly can act as such a linker.
+
+Containerizing the solution will introduce it as a service on a Linux node, fitting well with OpenShift and K8S model.
+
+### Configuration Examples
+In this section, several configuration examples are given using Ansible, showing how these are represented using the data model, and finally translated to NetworkManager configuration.
+
+See Appendix C
+
+
+## MILESTONES
+Phasing recommendations:
+1. PoC: Implement a minimalistic nmstate, use the proposed schema and commands to run it all from Ansible with the unified network modules.
+
+User Story: Setup an IPv4 address on a bond with two slaves, supporting the concept of transaction.
+
+2. Introduce basic configuration capabilities for bonds, bridges, vlans and utilize them through Ansible.
+3. Complement missing RHEL networking properties, focusing on properties that are needed by existing projects that use RHEL networking (RHV host agent, Cockpit). During this phase, RHV can be used, replacing its own host agent networking module.
+
+
+## Appendix A (Schema proposal)
+The schema presented in this section is an informal version, skipping the data types. It attempts to describe the general idea.
+```
+    +--rw interfaces
+        +--rw interface*
+           +--rw name
+           +--rw description
+           +--rw type (including vlan, bond, team, linux-bridge, ovs)
+           +--rw state
+           +--ro admin-status
+           +---ro link-status (up|down)
+           +---ro link-speed (base link speed for virtual interfaces)
+           +-- ethernet:
+              +-- rw auto-negotiation (yes|no)
+              +-- rw duplex (ro with auto-negotiation)
+              +-- rw speed (ro with auto-negotiation)
+              +-- rw flow-control (ro with auto-negotiation)
+           +--rw mac-address
+           +--rw vlan-id (for vlan type)
```

Comment:
In order to be consistent with the other type specific interfaces, these two properties (`vlan-id` and `vlan-base`) should be placed under a `vlan` subtree.
```
+--vlan:
   +-- id
   +-- base-iface

Reply:
Done